yt_idefix/data_structures.py

```diff
@@ -430,6 +458,270 @@ def _get_code_version(self) -> str:
+                self._default_definitions_header,
+            )
+            return
+
+        with open(self._definitions_header) as fh:
+            body = fh.read()
+        lines = C_io.strip_comments(body).split("\n")
+
+        for line in lines:
+            if (geom_match := re.fullmatch(_DEF_GEOMETRY_REGEXP, line)) is not None:
+                self.parameters["definitions"]["geometry"] = geom_match.group(1).lower()
+            elif (unit_match := re.fullmatch(_DEF_UNIT_REGEXP, line)) is not None:
+                unit = unit_match.group(1).lower() + "_unit"
+                expr = unit_match.group(2)
+                # Before evaluating the expression, replace the input parameters,
+                # pre-defined constants, code units and arithmetic operators
+                # that cannot be resolved. The order doesn't matter.
+                expr = re.sub(r"g_inputParam\[(\w+)\]", self._get_input_parameter, expr)
+                expr = re.sub(r"CONST_\w+", self._get_constants, expr)
+                expr = re.sub(r"UNIT_(\w+)", self._get_unit, expr)
+                expr = re.sub(r"sqrt", "np.sqrt", expr)
+                expr = re.sub(r"log", "np.log", expr)
+                self.parameters["definitions"][unit] = eval(expr)
+
+    def _get_input_parameter(self, match: re.Match) -> str:
+        """Replace matched input parameters with its value"""
+        key = match.group(1)
+        return str(self.parameters["Parameters"][key])
+
+    def _get_unit(self, match: re.Match) -> str:
+        """Replace matched unit with its value"""
+        key = match.group(1).lower() + "_unit"
+        return str(self.parameters["definitions"].get(key, 1.0))
+
+    def _get_constants(self, match: re.Match) -> str:
+        """Replace matched constant string with its value"""
+        key = match.group()
+        return str(pluto_def_constants[key])
+
+    def _get_code_version(self) -> str:
+        # take the first line of the header
+        # Code version is read from grid.out
+        """
+        In grid.out
+        # ***********
+        makrs the beginning and the end of header information
+        """
+        if not (hasattr(self, "grid_file")):
+            self.grid_file = os.path.join(
+                self.directory, "grid.out"
+            )  # data-loc/grid.out
+        if not (os.path.exists(self.grid_file)):
+            warnings.warn(
+                f"Could not determine code version from file {self.grid_file}"
+            )
+            return "unknown"
+
+        count = 0
+        version = ""
+
+        with open(self.grid_file) as fh:
+            txt = fh.readlines()
+
+            for _start, line in enumerate(txt):
+                if "# ***********" in line:
+                    count += 1
+                if count == 1:
+                    break
+            _start += 1  # This is the first line of the header info lines in grid.out (version info is here)
+            version = txt[_start].split()[2]
```

Comment:
Even better would be to match a regular expression, but here's a much simplified version of what you're doing here
```suggestion
        with open(self.grid_file) as fh:
            while not (line := fh.readline()).startswith("# ***********"):
                count += 1
            version = txt[count + 1].split()[2]
```

Reply:
I have incorporated your suggestions on this.